src/main/java/frc/robot/subsystems/swerve/SwerveConstants.java

```diff
@@ -0,0 +1,117 @@
+package frc.robot.subsystems.swerve;
+
+import com.ctre.phoenix6.configs.CANcoderConfiguration;
+import com.ctre.phoenix6.configs.TalonFXConfiguration;
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.math.Matrix;
+import edu.wpi.first.math.Nat;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.numbers.N1;
+import edu.wpi.first.math.numbers.N3;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.Alert;
+import edu.wpi.first.wpilibj.Alert.AlertType;
+import frc.robot.subsystems.swerve.Module.ModuleConstants;
+import frc.robot.subsystems.vision.Vision.VisionConstants;
+
+/**
+ * A template for constants for a swerve drivetrain. This template presumes rectangular modules that
+ * are symmetric around the center of the robot, talonfxs + cancoders, and that all modules have the
+ * same gear ratios Subclass this to make constants for a specific robot.
+ */
+public abstract class SwerveConstants {
```

Comment:
could this be a record or do we not really care

Reply:
record syntax is uglier imo, and this lets us do defaults which idt records can?